original file: [link to web](https://archive.ics.uci.edu/ml/datasets/Heart+Disease)

### first libraries

In [1]:
import pandas as pd
import numpy as np
import csv
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

### Loading Data

In [2]:
FeaturesCompletas=[
    'id','ccf','age','sex','painloc','painexer','relrest',
    'pncaden','cp','trestbps','htn','chol','smoke','cigs','years',
    'fbs','dm','famhist','restecg','ekgmo','ekgday','ekgyr','dig',
    'prop','nitr','pro','diuretic','proto','thaldur','thaltime','met',
    'thalach','thalrest','tpeakbps','tpeakbpd','dummy','trestbpd', 'exang','xhypo',
    'oldpeak', 'slope', 'rldv5','rldv5e','ca','restckm','exerckm','restef',
    'restwm', 'exeref','exerwm','thal','thalsev','thalpul','earlobe','cmo',
    'cday', 'cyr','num','lmt','ladprox','laddist','diag','cxmain',
    'ramus','om1','om2','rcaprox','rcadist','lvx1','lvx2','lvx3',
    'lvx4','lvf','cathef','junk','name']

In [3]:
# displaying all rows and columns
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [4]:
# first read of the dataset
arquivo = ('./data/cleveland.data')
lineList = [line.rstrip('\n') for line in open(arquivo, encoding = 'latin1')]
blocos = pd.Series(lineList)

In [6]:
# displaying data
blocos.head(30)

0            1 0 63 1 -9 -9 -9
1      -9 1 145 1 233 -9 50 20
2            1 -9 1 2 2 3 81 0
3          0 0 0 0 1 10.5 6 13
4     150 60 190 90 145 85 0 0
5      2.3 3 -9 172 0 -9 -9 -9
6        -9 -9 -9 6 -9 -9 -9 2
7           16 81 0 1 1 1 -9 1
8            -9 1 -9 1 1 1 1 1
9               1 1 -9 -9 name
10           2 0 67 1 -9 -9 -9
11     -9 4 160 1 286 -9 40 40
12           0 -9 1 2 3 5 81 0
13          1 0 0 0 1 9.5 6 13
14    108 64 160 90 160 90 1 0
15     1.5 2 -9 185 3 -9 -9 -9
16       -9 -9 -9 3 -9 -9 -9 2
17           5 81 2 1 2 2 -9 2
18           -9 1 -9 1 1 1 1 1
19              1 1 -9 -9 name
20           3 0 67 1 -9 -9 -9
21     -9 4 120 1 229 -9 20 35
22          0 -9 1 2 2 19 81 0
23          1 0 0 0 1 8.5 6 10
24    129 78 140 80 120 80 1 0
25     2.6 2 -9 150 2 -9 -9 -9
26       -9 -9 -9 7 -9 -9 -9 2
27          20 81 1 1 1 1 -9 1
28           -9 1 -9 2 2 1 1 1
29              7 3 -9 -9 name
dtype: object

In [7]:
# cleaning data after person number 298
blocos = blocos[:2820]

In [9]:
# displaying last lines
blocos.tail(20)

2800         297 0 57 1 -9 -9 -9
2801     -9 4 110 1 335 -9 20 15
2802           0 -9 1 0 5 3 84 0
2803             0 0 0 0 1 7 3 8
2804    143 62 130 78 110 80 1 0
2805        3 2 -9 91 1 -9 -9 -9
2806       -9 -9 -9 7 -9 -9 -9 5
2807           4 84 2 1 1 1 -9 1
2808           -9 2 -9 1 2 1 1 1
2809              1 1 -9 -9 name
2810         298 0 47 1 -9 -9 -9
2811       -9 3 130 1 253 -9 0 0
2812          0 -9 1 0 5 23 84 0
2813         0 0 0 0 1 9.5 -9 11
2814    179 65 170 82 130 80 0 0
2815        0 1 -9 55 0 -9 -9 -9
2816       -9 -9 -9 3 -9 -9 -9 5
2817          24 84 0 1 1 1 -9 1
2818           -9 1 -9 1 1 1 1 1
2819              1 1 -9 -9 name
dtype: object

In [11]:
# slicing following dataset's pattern 
a = blocos[0::10].str.split(expand=True).reset_index()
b = blocos[1::10].str.split(expand=True).reset_index()
c = blocos[2::10].str.split(expand=True).reset_index()
d = blocos[3::10].str.split(expand=True).reset_index()
e = blocos[4::10].str.split(expand=True).reset_index()
f = blocos[5::10].str.split(expand=True).reset_index()
g = blocos[6::10].str.split(expand=True).reset_index()
h = blocos[7::10].str.split(expand=True).reset_index()
i = blocos[8::10].str.split(expand=True).reset_index()
j = blocos[9::10].str.split(expand=True).reset_index()

In [12]:
# naming following dataset's pattern 
a.columns = ['indexx','id','ccf','age','sex','painloc','painexer','relrest']
b.columns = ['indexx','pncaden','cp','trestbps','htn','chol','smoke','cigs','years']
c.columns = ['indexx','fbs','dm','famhist','restecg','ekgmo','ekgday','ekgyr','dig']
d.columns = ['indexx','prop','nitr','pro','diuretic','proto','thaldur','thaltime','met']
e.columns = ['indexx','thalach','thalrest','tpeakbps','tpeakbpd','dummy','trestbpd', 'exang','xhypo']
f.columns = ['indexx','oldpeak', 'slope', 'rldv5','rldv5e','ca','restckm','exerckm','restef']
g.columns = ['indexx','restwm', 'exeref','exerwm','thal','thalsev','thalpul','earlobe','cmo']
h.columns = ['indexx','cday', 'cyr','num','lmt','ladprox','laddist','diag','cxmain']
i.columns = ['indexx','ramus','om1','om2','rcaprox','rcadist','lvx1','lvx2','lvx3']
j.columns = ['indexx','lvx4','lvf','cathef','junk','name']

In [14]:
# bringing all togueter
tabCLV = pd.concat([a , b , c , d , e, f, g, h, i, j] , axis = 1 ).drop('indexx' , axis=1)

In [15]:
tabCLV.head(10)

id ccf age sex painloc painexer relrest pncaden cp trestbps htn chol smoke  \
0   1   0  63   1      -9       -9      -9      -9  1      145   1  233    -9   
1   2   0  67   1      -9       -9      -9      -9  4      160   1  286    -9   
2   3   0  67   1      -9       -9      -9      -9  4      120   1  229    -9   
3   4   0  37   1      -9       -9      -9      -9  3      130   0  250    -9   
4   6   0  41   0      -9       -9      -9      -9  2      130   1  204    -9   
5   7   0  56   1      -9       -9      -9      -9  2      120   1  236    -9   
6   8   0  62   0      -9       -9      -9      -9  4      140   0  268    -9   
7   9   0  57   0      -9       -9      -9      -9  4      120   1  354    -9   
8  10   0  63   1      -9       -9      -9      -9  4      130   1  254    -9   
9  11   0  53   1      -9       -9      -9      -9  4      140   0  203    -9   

  cigs years fbs  dm famhist restecg ekgmo ekgday ekgyr dig prop nitr pro  \
0   50    20   1  -9       1       2     2      3    81   0    0    0   0   
1   40    40   0  -9       1       2     3      5    81   0    1    0   0   
2   20    35   0  -9       1       2     2     19    81   0    1    0   0   
3    0     0   0  -9       1       0     2     13    81   0    1    0   0   
4    0     0   0  -9       1       2     2      7    81   0    0    0   0   
5   20    20   0  -9       1       0     3      9    81   1    0    0   0   
6    0     0   0  -9       1       2     1     28    81   0    0    0   0   
7    0     0   0  -9       1       0     7     20    81   0    0    0   0   
8    0     0   0  -9       0       2     7      2    81   0    1    1   0   
9   20    25   1  -9       1       2     7      3    81   0    1    0   0   

  diuretic proto thaldur thaltime met thalach thalrest tpeakbps tpeakbpd  \
0        0     1    10.5        6  13     150       60      190       90   
1        0     1     9.5        6  13     108       64      160       90   
2        0     1     8.5        6  10     129       78      140       80   
3        0     1      13       13  17     187       84      195       68   
4        0     1       7       -9   9     172       71      160       74   
5        0     1    11.3       -9  16     178       73      165       70   
6        0     1       6        6   7     160       83      180       84   
7        0     1       9        6  10     163       84      165       80   
8        0     1       8        8   9     147       75      120       70   
9        1     1     5.5        3   7     155       86      185      120   

  dummy trestbpd exang xhypo oldpeak slope rldv5 rldv5e ca restckm exerckm  \
0   145       85     0     0     2.3     3    -9    172  0      -9      -9   
1   160       90     1     0     1.5     2    -9    185  3      -9      -9   
2   120       80     1     0     2.6     2    -9    150  2      -9      -9   
3   130       78     0     0     3.5     3    -9    167  0      -9      -9   
4   130       86     0     0     1.4     1    -9     40  0      -9      -9   
5   120       75     0     0     0.8     1    -9    127  0      -9      -9   
6   140      100     0     0     3.6     3    -9    122  2      -9      -9   
7   120       80     1     0     0.6     1    -9    122  0      -9      -9   
8   130      105     0     0     1.4     2    -9     75  1      -9      -9   
9   140      100     1     0     3.1     3    -9     68  0      -9      -9   

  restef restwm exeref exerwm thal thalsev thalpul earlobe cmo cday cyr num  \
0     -9     -9     -9     -9    6      -9      -9      -9   2   16  81   0   
1     -9     -9     -9     -9    3      -9      -9      -9   2    5  81   2   
2     -9     -9     -9     -9    7      -9      -9      -9   2   20  81   1   
3     -9     -9     -9     -9    3      -9      -9      -9   2    4  81   0   
4     -9     -9     -9     -9    3      -9      -9      -9   2   18  81   0   
5     -9     -9     -9     -9    3      -9      -9      -9   3   10  81   0   
6     -9     -

### understanding the dataset

Complete attribute documentation: <br>
1 id: patient identification number <br>
2 ccf: social security number (I replaced this with a dummy value of 0)  <br>
3 age: age in years  <br>
4 sex: sex (1 = male; 0 = female)  <br>
5 painloc: chest pain location (1 = substernal; 0 = otherwise)  <br>
6 painexer (1 = provoked by exertion; 0 = otherwise)  <br>
7 relrest (1 = relieved after rest; 0 = otherwise)  <br>
8 pncaden (sum of 5, 6, and 7)  <br>
9 cp: chest pain type  <br>
```
    -- Value 1: typical angina 
    -- Value 2: atypical angina 
    -- Value 3: non-anginal pain 
    -- Value 4: asymptomatic ```
10 trestbps: resting blood pressure (in mm Hg on admission to the hospital)  <br>
11 htn  <br>
`` 
Hypertensive heart disease refers to heart conditions caused by high blood pressure. The heart working under increased pressure causes some different heart disorders. Hypertensive heart disease includes heart failure, thickening of the heart muscle, coronary artery disease, and other conditions.``

12 chol: serum cholestoral in mg/dl  <br>
13 smoke: I believe this is 1 = yes; 0 = no (is or is not a smoker)  <br>
14 cigs (cigarettes per day)  <br>
15 years (number of years as a smoker)  <br>
16 fbs: (fasting blood sugar > 120 mg/dl) (1 = true; 0 = false)  <br>
17 dm (1 = history of diabetes; 0 = no such history)  <br>
18 famhist: family history of coronary artery disease (1 = yes; 0 = no)  <br>
19 restecg: resting electrocardiographic results  <br>
```
    -- Value 0: normal 
    -- Value 1: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV) 
    -- Value 2: showing probable or definite left ventricular hypertrophy by Estes' criteria ```
20 ekgmo (month of exercise ECG reading)  <br>
21 ekgday(day of exercise ECG reading)  <br>
22 ekgyr (year of exercise ECG reading)  <br>
23 dig (digitalis used furing exercise ECG: 1 = yes; 0 = no)  <br>
24 prop (Beta blocker used during exercise ECG: 1 = yes; 0 = no)  <br>
25 nitr (nitrates used during exercise ECG: 1 = yes; 0 = no)  <br>
26 pro (calcium channel blocker used during exercise ECG: 1 = yes; 0 = no)  <br>
27 diuretic (diuretic used used during exercise ECG: 1 = yes; 0 = no)  <br>
28 proto: exercise protocol  <br>
```
    1 = Bruce 
    2 = Kottus 
    3 = McHenry 
    4 = fast Balke 
    5 = Balke 
    6 = Noughton 
    7 = bike 150 kpa min/min (Not sure if "kpa min/min" is what was written!) 
    8 = bike 125 kpa min/min 
    9 = bike 100 kpa min/min 
    10 = bike 75 kpa min/min 
    11 = bike 50 kpa min/min 
    12 = arm ergometer ```
29 thaldur: duration of exercise test in minutes  <br>
30 thaltime: time when ST measure depression was noted  <br>
31 met: mets achieved  <br>
32 thalach: maximum heart rate achieved  <br>
33 thalrest: resting heart rate  <br>
34 tpeakbps: peak exercise blood pressure (first of 2 parts)  <br>
35 tpeakbpd: peak exercise blood pressure (second of 2 parts)  <br>
36 dummy  <br>
37 trestbpd: resting blood pressure  <br>
38 exang: exercise induced angina (1 = yes; 0 = no)  <br>
39 xhypo: (1 = yes; 0 = no)  <br>
40 oldpeak = ST depression induced by exercise relative to rest  <br>
41 slope: the slope of the peak exercise ST segment  <br>
```
    -- Value 1: upsloping 
    -- Value 2: flat 
    -- Value 3: downsloping ```
42 rldv5: height at rest  <br>
43 rldv5e: height at peak exercise  <br>
44 ca: number of major vessels (0-3) colored by flourosopy  <br>
45 restckm: irrelevant  <br>
46 exerckm: irrelevant  <br>
47 restef: rest raidonuclid (sp?) ejection fraction  <br>
48 restwm: rest wall (sp?) motion abnormality  <br>
```
    0 = none 
    1 = mild or moderate 
    2 = moderate or severe 
    3 = akinesis or dyskmem (sp?) ```
49 exeref: exercise radinalid (sp?) ejection fraction  <br>
50 exerwm: exercise wall (sp?) motion  <br>
51 thal: 3 = normal; 6 = fixed defect; 7 = reversable defect  <br>
52 thalsev: not used  <br>
53 thalpul: not used  <br>
54 earlobe: not used  <br>
55 cmo: month of cardiac cath (sp?) (perhaps "call")  <br>
56 cday: day of cardiac cath (sp?)  <br>
57 cyr: year of cardiac cath (sp?)  <br>
58 num: diagnosis of heart disease (angiographic disease status)  <br>
```
    -- Value 0: < 50% diameter narrowing 
    -- Value 1: > 50% diameter narrowing ```
(in any major vessel: attributes 59 through 68 are vessels) 
``
Blood vessel, a vessel in the human or animal body in which blood circulates. The vessels that carry blood away from the heart are called arteries, and their very small branches are arterioles.``
59 lmt  <br>
60 ladprox  <br>
61 laddist  <br>
62 diag  <br>
63 cxmain  <br>
64 ramus  <br>
65 om1  <br>
66 om2  <br>
67 rcaprox  <br>
68 rcadist  <br>
69 lvx1: not used  <br>
70 lvx2: not used  <br>
71 lvx3: not used  <br>
72 lvx4: not used  <br>
73 lvf: not used  <br>
74 cathef: not used  <br>
75 junk: not used  <br>
76 name: last name of patient (I replaced this with the dummy string "name") <br>

first clean due to the lack of info

1 id: patient identification number <br>
2 ccf: social security number (I replaced this with a dummy value of 0)  <br>
69 lvx1: not used  <br>
70 lvx2: not used  <br>
71 lvx3: not used  <br>
72 lvx4: not used  <br>
73 lvf: not used  <br>
74 cathef: not used  <br>
75 junk: not used  <br>
76 name: last name of patient (I replaced this with the dummy string "name") <br>
<br>
e também aquelas que não tem informação relevante, sendo seus valores iguais à -9

In [19]:
tabCLV.head(5)

age sex cp trestbps htn chol cigs years fbs famhist restecg dig prop nitr  \
0  63   1  1      145   1  233   50    20   1       1       2   0    0    0   
1  67   1  4      160   1  286   40    40   0       1       2   0    1    0   
2  67   1  4      120   1  229   20    35   0       1       2   0    1    0   
3  37   1  3      130   0  250    0     0   0       1       0   0    1    0   
4  41   0  2      130   1  204    0     0   0       1       2   0    0    0   

  pro diuretic thaldur thaltime met thalach thalrest tpeakbps tpeakbpd dummy  \
0   0        0    10.5        6  13     150       60      190       90   145   
1   0        0     9.5        6  13     108       64      160       90   160   
2   0        0     8.5        6  10     129       78      140       80   120   
3   0        0      13       13  17     187       84      195       68   130   
4   0        0       7       -9   9     172       71      160       74   130   

  trestbpd exang xhypo oldpeak slope rldv5e ca thal num lmt ladprox laddist  \
0       85     0     0     2.3     3    172  0    6   0   1       1       1   
1       90     1     0     1.5     2    185  3    3   2   1       2       2   
2       80     1     0     2.6     2    150  2    7   1   1       1       1   
3       78     0     0     3.5     3    167  0    3   0   1       1       1   
4       86     0     0     1.4     1     40  0    3   0   1       1       1   

  cxmain om1 rcaprox rcadist  
0      1   1       1       1  
1      2   1       1       1  
2      1   1       2       2  
3      1   1       1       1  
4      1   1       1       1

In [17]:
# other non-relevant features 
tabCLV.drop(['id' , 'ccf' , 'lvx1', 'lvx2', 'lvx3', 'lvx4', 'lvf', 'cathef', 'junk', 'name', 'pncaden', 'painloc', 'painexer', 'relrest', 'smoke', 'dm', 'proto', 'rldv5', 'restckm', 'exerckm', 'restef', 'restwm', 'exeref', 'exerwm', 'thalsev', 'thalpul', 'earlobe', 'diag', 'ramus', 'om2', 'ekgmo', 'ekgday', 'ekgyr', 'ekgmo', 'ekgday', 'ekgyr', 'cmo', 'cday' , 'cyr'] , axis=1 , inplace = True)

In [18]:
tabCLV.head(5)

age sex cp trestbps htn chol cigs years fbs famhist restecg dig prop nitr  \
0  63   1  1      145   1  233   50    20   1       1       2   0    0    0   
1  67   1  4      160   1  286   40    40   0       1       2   0    1    0   
2  67   1  4      120   1  229   20    35   0       1       2   0    1    0   
3  37   1  3      130   0  250    0     0   0       1       0   0    1    0   
4  41   0  2      130   1  204    0     0   0       1       2   0    0    0   

  pro diuretic thaldur thaltime met thalach thalrest tpeakbps tpeakbpd dummy  \
0   0        0    10.5        6  13     150       60      190       90   145   
1   0        0     9.5        6  13     108       64      160       90   160   
2   0        0     8.5        6  10     129       78      140       80   120   
3   0        0      13       13  17     187       84      195       68   130   
4   0        0       7       -9   9     172       71      160       74   130   

  trestbpd exang xhypo oldpeak slope rldv5e ca thal num lmt ladprox laddist  \
0       85     0     0     2.3     3    172  0    6   0   1       1       1   
1       90     1     0     1.5     2    185  3    3   2   1       2       2   
2       80     1     0     2.6     2    150  2    7   1   1       1       1   
3       78     0     0     3.5     3    167  0    3   0   1       1       1   
4       86     0     0     1.4     1     40  0    3   0   1       1       1   

  cxmain om1 rcaprox rcadist  
0      1   1       1       1  
1      2   1       1       1  
2      1   1       2       2  
3      1   1       1       1  
4      1   1       1       1

#### manually categorical mapping

In [20]:
def listagem (tabela):
    '''
    customized function to list unique values
    as the features goes by inside the dataset
    '''
    lista = []
    for x in tabela:
        y = x
        x = tabela[x].unique()
        print(y , x , '\n')

In [21]:
# transforming remaining strings-numbers into numbers 
tabCLV = tabCLV.apply(pd.to_numeric)

# listagem das variáveis
listagem(tabCLV)

age [63 67 37 41 56 62 57 53 44 52 48 54 49 64 58 60 50 66 43 40 69 59 42 55
 61 65 71 51 46 45 39 68 47 34 35 29 70 77 38 74 76] 

sex [1 0] 

cp [1 4 3 2] 

trestbps [145 160 120 130 140 172 150 110 132 117 135 112 105 124 125 142 128 170
 155 104 180 138 108 134 122 115 118 100 200  94 165 102 152 101 126 174
 148 178 158 192 129 144 123 136 146 106 156 154] 

htn [1 0] 

chol [233 286 229 250 204 236 268 354 254 203 192 294 256 263 199 168 239 275
 266 211 283 284 224 206 219 340 226 247 167 230 335 234 177 276 353 243
 225 302 212 330 175 417 197 198 290 253 172 273 213 305 216 304 188 282
 185 232 326 231 269 267 248 360 258 308 245 270 208 264 321 274 325 235
 257 164 141 252 255 201 222 260 182 303 265 309 307 249 186 341 183 407
 217 288 220 209 227 261 174 281 221 205 240 289 318 298 564 246 322 299
 300 293 277 214 207 223 160 394 184 315 409 244 195 196 126 313 259 200
 262 215 228 193 271 210 327 149 295 306 178 237 218 242 319 166 180 311
 278] 

cigs [50 40 20  0 75 30 1

##### analysing numerical features

In [23]:
# numericals
CLVnum = tabCLV[['age' , 'trestbps', 'chol', 'cigs', 'years', 'thaldur', 'thaltime', 'met', 'thalach', 'thalrest', 'tpeakbps', 'tpeakbpd', 'dummy', 'trestbpd', 'oldpeak', 'rldv5e', 'num']]

In [24]:
# descriptive statistics
CLVnum.describe()

age    trestbps        chol        cigs       years     thaldur  \
count  282.000000  282.000000  282.000000  282.000000  282.000000  282.000000   
mean    54.411348  131.563830  249.092199   16.460993   14.829787    8.418440   
std      9.053083   17.757496   51.217546   19.580300   15.564563    2.590569   
min     29.000000   94.000000  126.000000   -9.000000   -9.000000    1.800000   
25%     48.000000  120.000000  213.000000    0.000000    0.000000    6.500000   
50%     55.000000  130.000000  244.000000   10.000000   15.000000    8.500000   
75%     61.000000  140.000000  277.000000   30.000000   30.000000   10.075000   
max     77.000000  200.000000  564.000000   99.000000   54.000000   15.000000   

         thaltime         met     thalach    thalrest    tpeakbps    tpeakbpd  \
count  282.000000  282.000000  282.000000  282.000000  282.000000  282.000000   
mean     1.485816    9.753546  149.765957   75.117021  168.056738   78.744681   
std      6.693648    2.961870   22.923869   13.795809   23.575961   13.331206   
min     -9.000000    3.000000   71.000000   40.000000   84.000000   26.000000   
25%      0.000000    7.000000  133.250000   65.000000  154.000000   70.000000   
50%      3.000000    9.000000  153.500000   74.000000  168.000000   80.000000   
75%      6.000000   12.000000  165.750000   84.000000  184.000000   85.000000   
max     15.000000   18.000000  202.000000  119.000000  232.000000  120.000000   

            dummy    trestbpd     oldpeak      rldv5e         num  
count  282.000000  282.000000  282.000000  282.000000  282.000000  
mean   131.563830   84.946809    1.026950  123.585106    0.907801  
std     17.757496    9.477864    1.138825   46.783839    1.224894  
min     94.000000   50.000000    0.000000   24.000000    0.000000  
25%    120.000000   80.000000    0.000000   92.000000    0.000000  
50%    130.000000   85.000000    0.800000  118.000000    0.000000  
75%    140.000000   90.000000    1.600000  152.750000    2.000000  
max    200.000000  110.000000    6.200000  270.000000    4.000000

In [31]:
# replacing -9 to feature's mean's value
tabCLV['cigs'].replace(-9 , tabCLV['cigs'].mean() , inplace = True)
tabCLV['years'].replace(-9 , tabCLV['years'].mean() , inplace = True)
tabCLV['thaltime'].replace(-9 , tabCLV['thaltime'].mean() , inplace = True)

# setting yes or no to angina
tabCLV['num'] = tabCLV.num.replace({0:0 , 1:1, 2:1, 3:1, 4:1})

print('ok')

ok


In [32]:
# numerical tab
CLVnum = tabCLV[['age' , 'trestbps', 'chol', 'cigs', 'years', 'thaldur', 'thaltime', 'met', 'thalach', 'thalrest', 'tpeakbps', 'tpeakbpd', 'dummy', 'trestbpd', 'oldpeak', 'rldv5e', 'num']]

In [33]:
CLVnum.describe()

age    trestbps        chol        cigs       years     thaldur  \
count  282.000000  282.000000  282.000000  282.000000  282.000000  282.000000   
mean    54.411348  131.563830  249.092199   16.912429   15.252301    8.418440   
std      9.053083   17.757496   51.217546   19.278193   15.230634    2.590569   
min     29.000000   94.000000  126.000000    0.000000    0.000000    1.800000   
25%     48.000000  120.000000  213.000000    0.000000    0.000000    6.500000   
50%     55.000000  130.000000  244.000000   15.000000   15.000000    8.500000   
75%     61.000000  140.000000  277.000000   30.000000   30.000000   10.075000   
max     77.000000  200.000000  564.000000   99.000000   54.000000   15.000000   

         thaltime         met     thalach    thalrest    tpeakbps    tpeakbpd  \
count  282.000000  282.000000  282.000000  282.000000  282.000000  282.000000   
mean     4.051494    9.753546  149.765957   75.117021  168.056738   78.744681   
std      3.346613    2.961870   22.923869   13.795809   23.575961   13.331206   
min      0.000000    3.000000   71.000000   40.000000   84.000000   26.000000   
25%      1.485816    7.000000  133.250000   65.000000  154.000000   70.000000   
50%      3.000000    9.000000  153.500000   74.000000  168.000000   80.000000   
75%      6.000000   12.000000  165.750000   84.000000  184.000000   85.000000   
max     15.000000   18.000000  202.000000  119.000000  232.000000  120.000000   

            dummy    trestbpd     oldpeak      rldv5e         num  
count  282.000000  282.000000  282.000000  282.000000  282.000000  
mean   131.563830   84.946809    1.026950  123.585106    0.443262  
std     17.757496    9.477864    1.138825   46.783839    0.497654  
min     94.000000   50.000000    0.000000   24.000000    0.000000  
25%    120.000000   80.000000    0.000000   92.000000    0.000000  
50%    130.000000   85.000000    0.800000  118.000000    0.000000  
75%    140.000000   90.000000    1.600000  152.750000    1.000000  
max    200.000000  110.000000    6.200000  270.000000    1.000000

In [34]:
CLVnum.shape

(282, 17)

##### analysing categorical features

In [35]:
# categoricals
CLVcat = tabCLV[['sex', 'cp', 'htn', 'fbs' , 'famhist' , 'restecg', 'dig' , 'prop', 'nitr', 'pro', 'diuretic', 'exang', 'xhypo', 'slope', 'ca', 'thal',  'lmt', 'ladprox', 'laddist', 'cxmain', 'om1', 'rcaprox', 'rcadist']]

In [36]:
listagem(CLVcat)

sex [1 0] 

cp [1 4 3 2] 

htn [1 0] 

fbs [1 0] 

famhist [1 0] 

restecg [2 0 1] 

dig [ 0  1 -9] 

prop [ 0  1 -9] 

nitr [ 0  1 -9] 

pro [ 0  1 -9] 

diuretic [ 0  1 -9] 

exang [0 1] 

xhypo [0 1] 

slope [3 2 1] 

ca [ 0  3  2  1 -9] 

thal [ 6  3  7 -9] 

lmt [1 2] 

ladprox [1 2] 

laddist [1 2] 

cxmain [1 2] 

om1 [1 2] 

rcaprox [1 2] 

rcadist [1 2] 



In [37]:
# replacing -9 to feature's median's value
tabCLV['dig'].replace(-9 , tabCLV['dig'].median() , inplace = True)
tabCLV['prop'].replace(-9 , tabCLV['prop'].median() , inplace = True)
tabCLV['nitr'].replace(-9 , tabCLV['nitr'].median() , inplace = True)
tabCLV['pro'].replace(-9 , tabCLV['pro'].median() , inplace = True)
tabCLV['diuretic'].replace(-9 , tabCLV['diuretic'].median() , inplace = True)
tabCLV['ca'].replace(-9 , tabCLV['ca'].median() , inplace = True)
tabCLV['thal'].replace(-9 , tabCLV['thal'].median() , inplace = True)

print('ok')

ok


In [38]:
#tabela categórica
CLVcat = tabCLV[['sex', 'cp', 'htn', 'fbs' , 'famhist' , 'restecg', 'dig' , 'prop', 'nitr', 'pro', 'diuretic', 'exang', 'xhypo', 'slope', 'ca', 'thal',  'lmt', 'ladprox', 'laddist', 'cxmain', 'om1', 'rcaprox', 'rcadist']]

In [39]:
listagem(CLVcat)

sex [1 0] 

cp [1 4 3 2] 

htn [1 0] 

fbs [1 0] 

famhist [1 0] 

restecg [2 0 1] 

dig [0 1] 

prop [0 1] 

nitr [0 1] 

pro [0 1] 

diuretic [0 1] 

exang [0 1] 

xhypo [0 1] 

slope [3 2 1] 

ca [0 3 2 1] 

thal [6 3 7] 

lmt [1 2] 

ladprox [1 2] 

laddist [1 2] 

cxmain [1 2] 

om1 [1 2] 

rcaprox [1 2] 

rcadist [1 2] 



In [40]:
CLVcat.shape

(282, 23)

In [41]:
# getting the dummies for the categoricals
CLVcat2 = pd.concat([pd.get_dummies(tabCLV[col]) for col in CLVcat], axis=1, keys=CLVcat.columns)

In [42]:
CLVcat2.shape

(282, 53)

In [43]:
CLVcat2.describe()

sex                      cp                                      \
                0           1           1           2           3           4   
count  282.000000  282.000000  282.000000  282.000000  282.000000  282.000000   
mean     0.322695    0.677305    0.078014    0.152482    0.297872    0.471631   
std      0.468338    0.468338    0.268671    0.360127    0.458136    0.500082   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
50%      0.000000    1.000000    0.000000    0.000000    0.000000    0.000000   
75%      1.000000    1.000000    0.000000    0.000000    1.000000    1.000000   
max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   

              htn                     fbs                 famhist              \
                0           1           0           1           0           1   
count  282.000000  282.000000  282.000000  282.000000  282.000000  282.000000   
mean     0.382979    0.617021    0.851064    0.148936    0.379433    0.620567   
std      0.486977    0.486977    0.356658    0.356658    0.486108    0.486108   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000000    1.000000    0.000000    0.000000    0.000000   
50%      0.000000    1.000000    1.000000    0.000000    0.000000    1.000000   
75%      1.000000    1.000000    1.000000    0.000000    1.000000    1.000000   
max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   

          restecg                                 dig                    prop  \
                0           1           2           0           1           0   
count  282.000000  282.000000  282.000000  282.000000  282.000000  282.000000   
mean     0.489362    0.007092    0.503546    0.968085    0.031915    0.666667   
std      0.500776    0.084065    0.500876    0.176086    0.176086    0.472243   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000000    0.000000    1.000000    0.000000    0.000000   
50%      0.000000    0.000000    1.000000    1.000000    0.000000    1.000000   
75%      1.000000    0.000000    1.000000    1.000000    0.000000    1.000000   
max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   

                         nitr                     pro                diuretic  \
                1           0           1           0           1           0   
count  282.000000  282.000000  282.000000  282.000000  282.000000  282.000000   
mean     0.333333    0.755319    0.244681    0.900709    0.099291    0.886525   
std      0.472243    0.430662    0.430662    0.299584    0.299584    0.317737   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    1.000000    0.000000    1.000000    0.000000    1.000000   
50%      0.000000    1.000000    0.000000    1.000000    0.000000    1.000000   
75%      1.000000    1.000000    0.000000    1.000000    0.000000    1.000000   
max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   

                        exang                   xhypo                   slope  \
                1           0           1           0           1           1   
count  282.000000  282.000000  282.000000  282.000000  282.000000  282.000000   
mean     0.113475    0.673759    0.326241    0.978723    0.021277    0.478723   
std      0.317737    0.469670    0.469670    0.144561    0.144561    0.500435   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000000    0.000000    1.000000    0.000000    0.000000   
50%      0.000000    1.000000    0.000000    1.000000    0.000000    0.000000   
75%      0.000000    1.000000    1.000000    1.000000    0.000000    1.000000   
max      1.000000    1.000000    1.000000

##### concatenating tables

In [44]:
CLVfinal = pd.concat([CLVnum , CLVcat2] , axis = 1 )

In [45]:
CLVfinal.shape

(282, 70)

In [46]:
CLVfinal.describe()

age    trestbps        chol        cigs       years     thaldur  \
count  282.000000  282.000000  282.000000  282.000000  282.000000  282.000000   
mean    54.411348  131.563830  249.092199   16.912429   15.252301    8.418440   
std      9.053083   17.757496   51.217546   19.278193   15.230634    2.590569   
min     29.000000   94.000000  126.000000    0.000000    0.000000    1.800000   
25%     48.000000  120.000000  213.000000    0.000000    0.000000    6.500000   
50%     55.000000  130.000000  244.000000   15.000000   15.000000    8.500000   
75%     61.000000  140.000000  277.000000   30.000000   30.000000   10.075000   
max     77.000000  200.000000  564.000000   99.000000   54.000000   15.000000   

         thaltime         met     thalach    thalrest    tpeakbps    tpeakbpd  \
count  282.000000  282.000000  282.000000  282.000000  282.000000  282.000000   
mean     4.051494    9.753546  149.765957   75.117021  168.056738   78.744681   
std      3.346613    2.961870   22.923869   13.795809   23.575961   13.331206   
min      0.000000    3.000000   71.000000   40.000000   84.000000   26.000000   
25%      1.485816    7.000000  133.250000   65.000000  154.000000   70.000000   
50%      3.000000    9.000000  153.500000   74.000000  168.000000   80.000000   
75%      6.000000   12.000000  165.750000   84.000000  184.000000   85.000000   
max     15.000000   18.000000  202.000000  119.000000  232.000000  120.000000   

            dummy    trestbpd     oldpeak      rldv5e         num    (sex, 0)  \
count  282.000000  282.000000  282.000000  282.000000  282.000000  282.000000   
mean   131.563830   84.946809    1.026950  123.585106    0.443262    0.322695   
std     17.757496    9.477864    1.138825   46.783839    0.497654    0.468338   
min     94.000000   50.000000    0.000000   24.000000    0.000000    0.000000   
25%    120.000000   80.000000    0.000000   92.000000    0.000000    0.000000   
50%    130.000000   85.000000    0.800000  118.000000    0.000000    0.000000   
75%    140.000000   90.000000    1.600000  152.750000    1.000000    1.000000   
max    200.000000  110.000000    6.200000  270.000000    1.000000    1.000000   

         (sex, 1)     (cp, 1)     (cp, 2)     (cp, 3)     (cp, 4)    (htn, 0)  \
count  282.000000  282.000000  282.000000  282.000000  282.000000  282.000000   
mean     0.677305    0.078014    0.152482    0.297872    0.471631    0.382979   
std      0.468338    0.268671    0.360127    0.458136    0.500082    0.486977   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
50%      1.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
75%      1.000000    0.000000    0.000000    1.000000    1.000000    1.000000   
max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   

         (htn, 1)    (fbs, 0)    (fbs, 1)  (famhist, 0)  (famhist, 1)  \
count  282.000000  282.000000  282.000000    282.000000    282.000000   
mean     0.617021    0.851064    0.148936      0.379433      0.620567   
std      0.486977    0.356658    0.356658      0.486108      0.486108   
min      0.000000    0.000000    0.000000      0.000000      0.000000   
25%      0.000000    1.000000    0.000000      0.000000      0.000000   
50%      1.000000    1.000000    0.000000      0.000000      1.000000   
75%      1.000000    1.000000    0.000000      1.000000      1.000000   
max      1.000000    1.000000    1.000000      1.000000      1.000000   

       (restecg, 0)  (restecg, 1)  (restecg, 2)    (dig, 0)    (dig, 1)  \
count    282.000000    282.000000    282.000000  282.000000  282.000000   
mean       0.489362      0.007092      0.503546    0.968085    0.031915   
std        0.500776      0.084065      0.500876    0.176086    0.176086   
min        0.000000      0.000000      0.000000    0.000000    0.000000   
25%        0.000000      0.000000      0.000000    1.000000  

# Forecasting Angina

### learing with TrainTest 80/20

In [47]:
#************** definição das features explicativas e explicada **************
            #numéricas e categóricas: RESULTA EM OVERFITTING
#X = CLVfinal.drop(['num'] , axis =1)
#y = CLVfinal['num']
            #numéricas 
X = CLVnum.drop(['num'] , axis =1)
y = CLVnum['num']
        
        
        
#************** parcionando os dados em teste e treino **************
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=42)



#************** padronização dos features para mesma escala **************
from sklearn.preprocessing import StandardScaler 
sc = StandardScaler() 
X_train = sc.fit_transform(X_train) 
X_test = sc.transform(X_test) 
        

#************** importando a matriz de confusão para comparar modelos **************
from sklearn import metrics

##### balancing the dataset

In [48]:
#CLVfinal.groupby('num').age.count()
CLVnum.groupby('num').age.count()

num
0    157
1    125
Name: age, dtype: int64

In [49]:
#CLV_0 = CLVfinal[CLVfinal.num == 0]
#CLV_1 = CLVfinal[CLVfinal.num != 0]
CLV_0 = CLVnum[CLVnum.num == 0]
CLV_1 = CLVnum[CLVnum.num != 0]
print('não: ', len(CLV_0) )
print('sim: ', len(CLV_1) )

não:  157
sim:  125


In [50]:
CLV_0 = CLV_0.sample(n=125 , replace = True)
Cleveland = pd.concat([CLV_0 , CLV_1] , ignore_index = True)
print('linhas qndo 0 angina >> ' , len(CLV_0) , '|||', 'shape >> ' , Cleveland.shape)

linhas qndo 0 angina >>  125 ||| shape >>  (250, 17)


##### testing models

In [51]:
# importe vários modelos (coloque mais para testar)
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.linear_model import LogisticRegression


classifiers = [
    KNeighborsClassifier(3),
    GaussianNB(),
    LogisticRegression(),
    SVC(),
    DecisionTreeClassifier(),
    RandomForestClassifier(),
    GradientBoostingClassifier()]
    

for clf in classifiers:
    clf.fit(X_train, y_train)
    name = clf.__class__.__name__
    
    print("="*30)
    print(name)
    
    print('****Results****')
    y_pred = clf.predict(X_test)
    print("Accuracy:", metrics.accuracy_score(y_test, y_pred))
    print("Precision:", metrics.precision_score(y_test, y_pred))
    print("Recall:", metrics.recall_score(y_test, y_pred))

KNeighborsClassifier
****Results****
Accuracy: 0.6842105263157895
Precision: 0.6785714285714286
Recall: 0.6785714285714286
GaussianNB
****Results****
Accuracy: 0.7017543859649122
Precision: 0.6666666666666666
Recall: 0.7857142857142857
LogisticRegression
****Results****
Accuracy: 0.7017543859649122
Precision: 0.7037037037037037
Recall: 0.6785714285714286
SVC
****Results****
Accuracy: 0.6842105263157895
Precision: 0.6785714285714286
Recall: 0.6785714285714286
DecisionTreeClassifier
****Results****
Accuracy: 0.6842105263157895
Precision: 0.6785714285714286
Recall: 0.6785714285714286
RandomForestClassifier
****Results****
Accuracy: 0.7017543859649122
Precision: 0.7619047619047619
Recall: 0.5714285714285714


C:\Users\cleci\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\cleci\Anaconda3\lib\site-packages\sklearn\ensemble\forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


GradientBoostingClassifier
****Results****
Accuracy: 0.7017543859649122
Precision: 0.7037037037037037
Recall: 0.6785714285714286


#### GaussianNB and LogisticRegression have shown good accuracy. Between then, the first has the highest recall while the second has better precision. '\n'
#### As long as this dataset is about health issues, type one error is more critical and the machine has to learn more efficiently. Hence, GaussianNB model is the best one to be worked on.